### ISSUE TITLE
CA1508 object?.Content == null is always 'false', when checking two different objects from same async function.

### ISSUE THREAD

**Issue by nicro950** [dotnet/roslyn-analyzers#5245]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.0-preview.6.21353.1](https://dotnet.microsoft.com/download/dotnet/6.0)

### Describe the bug
When using the following code, the analyzer complains that the obj2?.Content == null is always false, even if we set it to null. It looks like the analyzer thinks that obj == obj2 since it is returned by the same async function. If the async is removed, then the warning disappears.

```csharp
        static async Task Main(string[] args)
        {
            var obj = await GetSomeObject("something");
            if (obj?.Content == null)
                return;

            var obj2 = await GetSomeObject(null);
            if (obj2?.Content == null) // The warning is on this line saying obj2.Content == null is always false
                return;
        }

        static Task<SomeObject> GetSomeObject(string? content)
        {
            return Task.FromResult(new SomeObject() { Content = content });
        }
        public class SomeObject
        {
            public string? Content { get; set; }
        }
```

### Steps To Reproduce

1. Create new project, this was created with .net 5, but with dotnet 6 preview 6 installed
2. Enable project analyzers in csproj with:
```xml
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
```
3. Paste the code into the console app, and the code should warn about `== null` is always false.

### Expected behavior
Not reporting warning

### Actual behavior
False positive warning

**Comment by mavasani** [comment]
Verified that this repros in .NET 7 SDK, but no longer repros in the latest .NET 8 preview SDK
